### PR TITLE
mir::Server: Fix signal handling

### PR DESCRIPTION
**Pull request by RAOF**
Fixes: #2364, MirServer/wlcs#30

### CODE REVIEW

File: src/server/run_mir.cpp [hunk 1]
```diff
@@ -47,16 +48,126 @@ extern "C" void perform_emergency_cleanup()
...
+         * Re-raising the signal will destroy this context, so call the handler directly.
+         */
+        (*old_handler->sa_sigaction)(sig, info, ucontext);
+        /* We don't *expect* to get here - fatal signal handlers will *generally* re-raise the signal,
+         * and so don't return.
+         *
+         * However, we've just performed emergency cleanup, so if the handler was trying to fix things up
+         * and continue we unfortunately can't let them.
+         */
+        std::abort();
```
Comment (AlanGriffiths) [inline src/server/run_mir.cpp:147]:
```suggest
        mir::fatal_error_abort("Unsupported attempt to continue after a fatal signal: "s + signum_to_string(sig) );
```

File: src/server/run_mir.cpp [hunk 2]
```diff
@@ -47,16 +48,126 @@ extern "C" void perform_emergency_cleanup()
...
+    case SIGABRT:
+        return "SIGABRT";
+    case SIGFPE:
+        return "SIGFPE";
+    case SIGSEGV:
+        return "SIGSEGV";
+    case SIGBUS:
+        return "SIGBUS";
+    default:
+        return "(Unknown signal)";
```
Comment (AlanGriffiths) [inline src/server/run_mir.cpp:172]: Maybe give the signal number in this case?

File: src/server/run_mir.cpp [hunk 3]
```diff
@@ -110,7 +221,20 @@ void mir::run_mir(
...
+
+                    if (sigaction(sig, &sig_handler_desc, old_handlers.initialiser_for_signal(sig)))
+                    {
+                        using namespace std::string_literals;
+                        BOOST_THROW_EXCEPTION((
+                            std::system_error{
+                                errno,
+                                std::system_category(),
+                                "Failed to install signal handler for "s + signum_to_string(sig)
+                            }));
```
Comment (AlanGriffiths) [inline src/server/run_mir.cpp:239]: Is throwing an exception actually useful?

File: src/server/run_mir.cpp [hunk 4]
```diff
@@ -110,7 +221,20 @@ void mir::run_mir(
             {
                 for (auto sig : intercepted)
                 {
-                    old_handler[sig] = signal(sig, fatal_signal_cleanup);
+                    struct sigaction sig_handler_desc;
+                    sig_handler_desc.sa_flags = SA_SIGINFO;
+                    sig_handler_desc.sa_sigaction = &fatal_signal_cleanup;
+
+                    if (sigaction(sig, &sig_handler_desc, old_handlers.initialiser_for_signal(sig)))
```
Comment (AlanGriffiths) [inline src/server/run_mir.cpp:231]: This lifetime of temporary trick with `AtomicOutPtr` is quite cool.

**Review by AlanGriffiths — APPROVED** [review]
This does what it claims. It is unfortunate we need to special case SIGBUS because of non-local knowledge (of libwayland internals).

I think the error handling could be simplified, but not blocking on that.

bors delegate

File: src/server/run_mir.cpp (hunk 4 above)
Comment (RAOF) [reply src/server/run_mir.cpp:231]: Yeah! It was really interesting seeing the summary of P1132R8 scroll across my twitter just before I had a need for such a gadget!

File: src/server/run_mir.cpp (hunk 3 above)
Comment (RAOF) [reply src/server/run_mir.cpp:239]: As opposed to simply `fatal_error_abort()`? Unlike the usages above, here we're in a context where there's going to be a catch handler further up the stack, and throwing gets us a nicer error and lets consuming code do something before dying.

If the alternative is not checking the return of `sigaction()` then that's maybe ok; if I understand it correctly that call can only fail in cases where we've made a programmer error. On the other hand, I *like* getting told when I've made a programming error ;)

**Comment by RAOF** [comment]
bors r=AlanGriffiths

**Review by AlanGriffiths — APPROVED** [review]
bors r+

**Comment by AlanGriffiths** [comment]
bors retry

**Pull request merged**